src/main/java/io/cdap/plugin/servicenow/sink/service/ServiceNowSinkAPIRequestImpl.java

```diff
@@ -0,0 +1,267 @@
+
+import org.apache.http.Header;
+import org.apache.http.HttpHeaders;
+import org.apache.http.HttpStatus;
+import org.apache.http.entity.ContentType;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.message.BasicHeader;
+import org.apache.oltu.oauth2.common.exception.OAuthProblemException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import javax.ws.rs.core.MediaType;
+
+/**
+ * Implementation class for ServiceNow Batch Rest API.
+ */
+public class ServiceNowSinkAPIRequestImpl {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ServiceNowSinkAPIRequestImpl.class);
+
+  private static Integer counter = 1;
+  private static Integer batchRequestIdCounter = 1;
+  private final ServiceNowSinkConfig config;
+  private final ServiceNowTableAPIClientImpl restApi;
+  private final Gson gson = new Gson();
+  private final JsonParser jsonParser = new JsonParser();
+  private boolean isCreated;
+
+
+  public ServiceNowSinkAPIRequestImpl(ServiceNowSinkConfig conf) {
+    this.config = conf;
+    restApi = new ServiceNowTableAPIClientImpl(config);
+  }
+
+  public RestRequest getRestRequest(JsonObject jsonObject) {
+    JsonElement jsonElement = jsonObject.get(ServiceNowConstants.SYS_ID);
+    String sysId = null;
+    // sys_id is mandatory for update operation
+    if (jsonElement == null && config.getOperation().equals(ServiceNowConstants.UPDATE_OPERATION)) {
+      throw new IllegalStateException("No sys_id found in the record to be updated");
+      // sys_id need to be removed for insert operation
+    } else if (jsonElement != null && config.getOperation().equals(ServiceNowConstants.INSERT_OPERATION)) {
+      jsonObject.remove(ServiceNowConstants.SYS_ID);
+      // this condition will run when jsonElement is not null and operation is update
+    } else if (jsonElement != null) {
+      sysId = jsonElement.getAsString();
+    }
+    String data = jsonObject.toString();
+    String encodedData = Base64.getEncoder().encodeToString(data.getBytes(StandardCharsets.UTF_8));
+
+    RestRequest restRequest = new RestRequest();
+    restRequest.setUrl(getUrl(sysId));
+    restRequest.setId(counter.toString());
+    // Increment Rest Request counter
+    counter++;
+    restRequest.setHeaders(getHeaders());
+    restRequest.setMethod(ServiceNowConstants.HTTP_POST);
+    if (config.getOperation().equals(ServiceNowConstants.UPDATE_OPERATION)) {
+      restRequest.setMethod(ServiceNowConstants.HTTP_PUT);
+    }
+    restRequest.setBody(encodedData);
+    return restRequest;
+  }
+
+  /**
+   * Inserts/Updates the list of records into ServiceNow table
+   *
+   * @param restRequestList The list of restRequests
```

Comment:
It's not clear what is the difference between restRequestList and records?

Reply:
Both are the similar actually but differ when the method: `createPostRequest()` gets called from `retryUnservicedRequests()`. I had to use two different variables because while implementing the retry logic for calling the API for unserviced requests, I needed the list of all the rest requests in a single batch as well as the list of unserviced requests to pass as arguments in the method: `retryUnservicedRequests()`

This method: `retryUnservicedRequests()` calls the method `createPostRequest()` every time there is a batch of requests which are left unserviced.
It takes three arguments:

1. `restRequestList`, The total number of requests in that batch
2. `records`, The total number of unserviced requests which would be retried and,
3. `accessToken`, The access token

The `restRequestList` and `records` have the same number of requests only for the first time it gets called from `createPostRequestRetryableMode()` method.

Reply:
hmm, I still don't understand the need for 2 request lists. Why can't we do something like this:
```
public void createPostRequest(List<RestRequest> restRequestList, String accessToken) {

    make POST request with ids in restRequestList
    if (response contains unserviced requests) {
       createPostRequest(unservicedRequests, accessToken);
}
```

Reply:
We need two variables because the unserviced request array just contains the request ids. These ids are sorted and used as an index while fetching records from the original batch. 

I've separated out the two methods 326fadb444e9fa01ddc710c3ae549590703f264e:

1. `processUnservicedRequestsArray()` : Converts the unserviced request array into unserviced requests list.
2. `createPostRequest()` : Calls the ServiceNow Batch API again with the batch of unserviced requests.

Reply:
records should be renamed as unservicedRequests